### ISSUE TITLE
Proposal: Taking address of temporary value is disallowed.

### ISSUE THREAD

**Issue by JesseRMeyer** [ziglang/zig#4355]
Both code examples compile and satisfy the type system but have subtly different semantics.

```zig
var d12_device: *ID3D12Device = undefined;
var result = D3D12CreateDevice(null, D3D_FEATURE_LEVEL_12_1, &IID_ID3D12Device, @ptrCast(**c_void, &d12_device));
```

```zig
var d12_device: *ID3D12Device = undefined;
var result = D3D12CreateDevice(null, D3D_FEATURE_LEVEL_12_1, &IID_ID3D12Device, &@ptrCast(*c_void, d12_device));
```

@ptrCast returns a temporary (unnamed) variable, so its address is likewise transient.  I cannot think of a valuable use-case for taking its address, and have felt this as an (educational) footgun about how Zig works.  

Is there a high value use-case for taking the address of a temporary variable?

If not, I propose this should be considered an error.

**Comment by MasterQ32** [comment]
They main question that comes up is:

Is the following code undefined behaviour or does Zig guarantee that the expression value of `try footgunnery(1)` is alive until "end of sope"

```zig
const T = struct { val: i32 };

fn footgunnery(v: i32) !T {
    … // here be magic implementation
}

fn callme() !i32 {
    const ptr1 = &(try footgunnery(1)).val;
    const ptr2 = &(try footgunnery(2)).val;
    return ptr1.* + ptr2.*;
}
```

For function parameters there should be at least the rule that taking a _mutable reference_ of a temporary is forbidden as i cannot imagine when it is not a bug.

Allowing a _const reference_ to a temporary (even a literal) can sometimes come in handy when interfacing generic C apis:
```zig
extern fn setValue(type: MyTypeStruct, value: *const c_void) void;

fn foo() void {
    setValue(.i32, &as(i32, 42));
}
```

**Comment by JesseRMeyer** [comment]
Good point.  The real danger comes from taking the address of temporary mutable references.

**Comment by dbandstra** [comment]
Footgun:

```zig
var opt: ?u32 = 4;

var ptr = &(opt orelse return);

ptr.* = 2;

std.debug.warn("{}\n", .{ opt }); // prints 4
```

Whereas `var ptr = &opt.?;` works. I'm not sure why.

**Comment by dbandstra** [comment]
Linking #7663 since it's at least somewhat related (taking address of struct literals).

**Comment by mlugg** [comment]
No longer relevant because address-of temporaries now returns a const pointer, solving the footgun.
